infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,243 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use digest::Digest;
+use tari_crypto::keys::PublicKey;
+
+#[derive(Debug, Error)]
+pub enum NodeIdError {
+    OutOfBounds,
+}
+
+#[derive(Clone, Debug, Eq, PartialOrd, Ord)]
+pub struct NodeDistance([u8; 32]);
+
+impl NodeDistance {
+    /// Construct a new zero distance
+    pub fn new() -> NodeDistance {
+        (NodeDistance::from(&[0; 32]))
+    }
+
+    /// Construct a node distance from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeDistance {
+        (NodeDistance(*array))
```

Comment:
Wouldn't it be more flexible to `impl From<&[u8]> for NodeDistance` here? 

Pros:
* Works with `Vec<u8>`
* Longer Arrays / slices

Cons:
* None really, but you have to copy the data and do some length checks

Reply:
Doing it this way, the number of input bytes cannot be fixed to a set size, resulting in the From sometimes failing depending on the input. I changed it to a try_from ```impl TryFrom<&[u8]> for NodeDistance```

Reply:
Ah yes, that's what you want :)